core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java

```diff
@@ -194,184 +197,278 @@ public ClusterManagerStatus getStatus() {
    }
 
    @Override
-   public CacheStatusResponse handleJoin(String cacheName, Address joiner, CacheJoinInfo joinInfo,
-         int joinerViewId) throws Exception {
-      ClusterCacheStatus cacheStatus;
-      clusterManagerLock.lock();
-      try {
-         waitForJoinerView(joiner, joinerViewId, joinInfo.getTimeout());
+   public CompletionStage<CacheStatusResponse> handleJoin(String cacheName, Address joiner, CacheJoinInfo joinInfo,
+                                                          int joinerViewId) {
+      CompletionStage<Void> viewStage;
+      if (joinerViewId <= viewId && clusterManagerStatus != ClusterManagerStatus.RECOVERING_CLUSTER) {
+         viewStage = CompletableFutures.completedNull();
+      } else {
+         if (trace) {
+            log.tracef("Delaying join request from %s until view %s is installed (and cluster status is recovered)",
+                       joiner, joinerViewId);
+         }
+         viewStage = joinViewFuture.newConditionStage(ctmi -> ctmi.canHandleJoin(joinerViewId),
+                                                      joinInfo.getTimeout(), MILLISECONDS)
+                                   .exceptionally(e -> wrapJoinTimeoutException(joinerViewId, e));
+      }
 
+      // After we have the right view, obtain the ClusterCacheStatus
+      return viewStage.thenApply(v -> prepareJoin(cacheName, joiner, joinInfo, joinerViewId))
+                      .thenApply(cacheStatus -> {
```

Comment:
you could have done everything inside the same Function.

Reply:
Yep, I think I made `prepareJoin()` return a `CompletionStage` at some point but that's no longer the case.